### PR TITLE
Vectorize Convert.ToBase64String using SSSE3

### PR DESCRIPTION
**Pull request by EgorBo**
This PR improves `Convert.ToBase64String` performance using SSSE3 instructions. 
It's based on "[Base64 encoding with SIMD instructions](http://0x80.pl/notesen/2016-01-12-sse-base64-encoding.html)" article by Wojciech Muła

Benchmark:
```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;
using System.Collections.Generic;

namespace ConsoleApp143
{
    public class ToBase64StringBenchmarks
    {
        public static IEnumerable<object[]> TestDataForGraph()
        {
            var rand = new Random(314666); // fixed "seed"
            for (int i = 0; i < 100; i++)
            {
                var data = new byte[i];
                for (int j = 0; j < i; j++)
                    data[j] = (byte)rand.Next(0, byte.MaxValue);
                yield return new object[] { data, i };
            }
        }

        [Benchmark]
        [ArgumentsSource(nameof(TestDataForGraph))]
        public string ToBase64(byte[] testData, int inputSize /* argument for report */) =>
            Convert.ToBase64String(testData, Base64FormattingOptions.InsertLineBreaks);

        static unsafe void Main(string[] args) => 
            BenchmarkSwitcher.FromAssembly(typeof(ToBase64StringBenchmarks).Assembly).Run(args);
    }
}
```
### Windows 10.0.17134.523, Core i7-8700K 3.7GHz (Coffee Lake):
image

### macOS 10.13.6, Core i7-4980HQ 2.8GHz (Haswell):
image

SSSE3-based implementation is limited with `input.Length>36` condition in order to avoid regressions for smaller values (the best value for my Skylake, Coffee Lake and Haswell based machines).

### CODE REVIEW

**Comment by stephentoub** [comment]
> For smaller input arrays according to my benchmark, performance shows up after input.Length >= 50

The graph doesn't show below 24... is there a regression for small values?  (It's pretty common to use base-64 encoding with small values, such as in various HTTP headers.)

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: Might be more efficient if this was a readonly static (at least until we optimize the JIT to make this a JIT-time constant).

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: Might be worth checking the perf of the 128-bit equivalent...

Not only would it let us cover smaller cases, but it will likely be overall more performant (in real-world scenarios) for common inputs (which likely won't be aligned and won't execute enough iterations that the benefits of using 256-bit paths outweighs the drawbacks).

**Comment by gfoidl** [comment]
FYI: https://github.com/dotnet/corefx/issues/32365 (will do when I get some time for this) (and https://github.com/gfoidl/Base64)

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (gfoidl) [inline src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: You should hoist these AVX-constants before the loop, so they are kept in (ymm-) registers, instead of creating the in eacht iteration.

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (gfoidl) [inline src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: See https://github.com/gfoidl/Base64/blob/b28b92389197cb3e37c2eb07c6aa0018d109c8db/source/gfoidl.Base64/Internal/Vector256Helper.cs#L71 for a more performant way.

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: @tannergooding yeah I did it in some commit but it adds a static constructor to `Convert` (but probably there is already so it's fine).

**Comment by EgorBo** [comment]
@gfoidl oh, didn't see your work. I did this just to practice and test Intrinsics API 🙂

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: I did it initially but got even worse output (spills?) will check again later

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: Will check on Haswell and Skylake, this limit should be significantly lower for SSE.

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: yeah makes sense but need to implement "InsertBrakeLines" first

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: Thanks, looks nicer! will compare outputs :)

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: Added SSSE3 impl and updated the graph, according to it SSSE is better than AVX for data.Length <120

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: UPD: for my CoffeeLake-machine the limit for SSE is 38 elements :)

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (fiigii) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: > I did it initially but got even worse output (spills?)

Currently, `Create` is not yet optimized to cache constant vector by the compiler (I plan to make it in the next steps).
So, I suggest defining all the constant vector as `static` members.

`Create(T)` (one argument) may be an exception that can generate a single `broadcast` instruction on AVX2 CPUs.

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: nit: Might be nice to name this one `Ssse3`, since that is the requirement for it

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: Is this a case where, for the final iteration, you could do a back-track and mask the already processed elements?

Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: um not sure I understand what you mean - if i != inputLength it fully forwards data to `ConvertToBase64Array` to complete and does `return;` if `i == inputLength` but inputLength % 3 != 0 - we need to insert paddings - we can ask `ConvertToBase64Array` to do it (see AVX impl) or insert it here (for better perf?)

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (EgorBo) [inline src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: these longs are Vector<ulong> representation of https://github.com/WojciechMula/base64simd/blob/master/encode/lookup.avx2.cpp#L72-L80
will fix it

File: src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs
Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: This looks like its handling trailing elements...

In a number of vectorized algorithms, it is possible to handle the trailing elements still with vectorization. This is accomplished by backtracking so that the final read is exactly 16-bytes, and then masking out any elements that were already processed. For example, we do that with `Scale` (multiplying a matrix by a scalar value) here: https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.CpuMath/SseIntrinsics.cs#L473

I was wondering if this algorithm would allow the same, as it (the additional couple instructions for masking) is often cheaper than looping over the remaining elements in scalar.

Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: @tannergooding there can be a case when I have a final `Vector256<byte>` (32 bytes) and there is no room for it in the output string to do `Avx2.Store`. It can crash with `ExecutionEngineException`.
I tried to remove this trailing elements processing and do everything on vectors and it worked! but with a chance of the crash 🙂

Comment (gfoidl) [reply src/System.Private.CoreLib/shared/System/Convert.Base64.Avx2.cs]: > and then masking out any elements that were already processed

Masking isn't needed, as the algorithm is idempotent, so you can process the already processed elements again (yielding the same encoded results).

But in my tests the backtracking introduced more overhead (especially to handle padding in a correct way, with setting a pad-vector and or-ing the result), so that a simple loop for the final elements resulted in better performance.

**Comment by EgorBo** [comment]
@tannergooding @stephentoub @fiigii @gfoidl I updated the PR and its description (added graphs). Could you please take a look?

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (fiigii) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: Why `Ssse3.IsSupported`?

Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: @fiigii Ah, I see, Vector128.Create has a software fallback, will remove.

Comment (gfoidl) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: Would it be better to use `default` when `Ssse3.IsSupported` results in `false`? So the software fallback isn't executed, as it isn't needed anyway.

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (gfoidl) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: Returning a value tuple for larger inputs (`length >>`) maybe no (perf-) problem, but for small input lengths it could be better to use `out`, so that the JIT may use registers instead of the stack. Though here I'm not sure what the JIT emits, as there a quite a lot of arguments involved.

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: @gfoidl I guess the software fallback won't be executed anyway since the code that uses these fields will be eliminated if SSSE3 is not supported

Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: oh, actually no it will - Convert already has an implicit static cctor

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: `s_base64ShiftLut` is a static readonly field (constant) - so I guess it should be kept in a register, will check asm once again for loads in this place

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: I didn't notice any noticeable performance regressions after I added this block for any values

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: Example: https://gist.github.com/EgorBo/4f62e9da70f2eb6b61cc3334da709764
Asm diff: 
image

**Review by fiigii — APPROVED** [review]
LGTM.

**Comment by fiigii** [comment]
Recently, we changed all the platform-agnostic intrinsic from instance methods to extension methods. Please rebase and re-test to ensure no conflict.

**Comment by gfoidl** [comment]
Also check https://github.com/dotnet/coreclr/pull/21833#discussion_r255901694 before merging please.

**Comment by EgorBo** [comment]
@gfoidl just checked, on the left `s_base64ShiftLut` is a static readonly field.
On the right - a local variable.
image
I expected `s_base64ShiftLut` as a static readonly field to be loaded into some xmm11 before the loop 😕

**Comment by gfoidl** [comment]
> as a static readonly field to be loaded into some xmm11 before the loop

(Unfortunately) The JIT won't hoist this, so we need to help it. 
I don't think this is a JIT-bug here, rather a known feature (given the time constraints it has).

**Comment by EgorBo** [comment]
@gfoidl so are you suggesting to add
`Vector128<byte> localShiftLut = s_base64ShiftLut;`
before the loop?

**Comment by gfoidl** [comment]
Yeah, I would do so, as it can be kept in the register. But check if it doesn't get spilled.

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: nit: Would be nice to have the comment that you load: `[xxxx|WWWZ|ZZYY|YXXX]` and then shuffle it to `[WWWW|ZZZZ|YYYY|XXXX]`

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: and that the overall layout is: `[bbbbcccc|CCdddddd|aaaaaaBB|bbbbcccc]`

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: nit: the paper looks to differentiate between `BB` and `bbbb` as well as `CC` and `cccc`

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: Did you check if having the `BMI2` path was better? The paper seems to indicate 15% and it should be available on most modern processors.

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: nit: Might be nice to indicate which of the algorithms you are using from the paper. This looks to be the `"Single pshufb method"`

**Review by tannergooding — APPROVED** [review]
Overall, LGTM. Just a couple of questions/nits

**Comment by stephentoub** [comment]
@EgorBo, are you still working on this?

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: I copied the comments from the C implementation: https://github.com/WojciechMula/base64simd/blob/master/encode/encode.sse.cpp#L20-L59
(I've just added the comments for Load + Shuffle)

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: I am not sure I follow, the comments are 100% copied from https://github.com/WojciechMula/base64simd/blob/master/encode/encode.sse.cpp#L20-L59 🙂

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: > I am not sure I follow, the comments are 100% copied from 

Aren't we then missing the appropriate 3rd-party notice information, copying the relevant licensing information into this file, etc.?
cc: @richlander

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: At the end of the article there are performance results and it looks like BMI2-based impl is slower.
I'll try to benchmark it but the code will be bigger (especially due to new static fields) and if you don't mind I'll check the performance later, maybe as part of a new PR if this one is merged.

**Comment by EgorBo** [comment]
@stephentoub updated the comments.
I guess this PR intersects with @gfoidl https://github.com/dotnet/corefx/pull/34529 who started to work on this earlier (and my PR focuses only on Encoding).

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (EgorBo) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: updated the THIRD-PARTY-NOTICES.txt

**Comment by gfoidl** [comment]
@EgorBo I wouldn't call it "intersects", as the other PR is for span-based byte -> byte encoding / decoding, whilst this one is for byte -> string (with line-breaks). So similar, but different targets.

If there would be no need for line-breaks, so the base64 encoding in `Convert` could be based on `System.Buffers.Text.Base64`.

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: > I am not sure I follow, the comments are 100% copied from 

I was commenting that the paper, in some places, differentiates between uppercase `BB` and lowercase `bbbb`. The source code doesn't seem to have maintained that in all the various places.

For example, the paper calls out:
> Input to this step are 32-bit words, each having following layout:
> ```
> [bbbbcccc|CCdddddd|aaaaaaBB|bbbbcccc]
> ```
> where bits aaaaaa, BBbbbb, ccccCC and dddddd are 6-bit indices. The output of this step has to be:
> ```
> [00dddddd|00ccccCC|00BBbbbb|00aaaaaa]
> ```

This looks to be particularly important for tracking which bits flow where

**Comment by danmosemsft** [comment]
Resolved merge conflict so we can get test results.

**Comment by danmosemsft** [comment]
@tannergooding if tests pass is this ready to merge?

**Comment by tannergooding** [comment]
I'll give this one more pass after lunch.

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: This was renamed to `StoreScalar` for the `ulong*` overload

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: CC. @EgorBo, I think this is the only thing that needs to be fixed and then this can be merged after tests pass.

**Comment by sandreenko** [comment]
@EgorBo do you think that PR can be finished before the consolidation (in next 2 weeks)?

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: A short comment describing each constant would be useful.

It's also not clear why these are static readonly, but several of the others (such as `tt0-tt8`) are not

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: It would still be nice to see this addressed.

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: Any reason this isn't a static local function? Inlining?

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (tannergooding) [inline src/System.Private.CoreLib/shared/System/Convert.cs]: Having the side effect only hit if `insertLineBreaks` is true, but required for both the true and false scenarios is non-obvious.

It would be nice to move the `charCount += 16` out separately

File: src/System.Private.CoreLib/shared/System/Convert.cs
Comment (saucecontrol) [reply src/System.Private.CoreLib/shared/System/Convert.cs]: Given https://github.com/dotnet/coreclr/issues/17225 and https://github.com/dotnet/coreclr/issues/26976, it would be more efficient processing and space-wise to use the `ROS<byte>` read-only property trick on these, especially since they're only used by code behind a `Ssse3.IsSupported` check.

**Comment by saucecontrol** [comment]
> SSSE3-based implementation is limited with input.Length>36 condition in order to avoid regressions for smaller values (the best value for my Skylake, Coffee Lake and Haswell based machines).

Is the 36-byte cutover point appropriate for 32-bit as well?  There are more than 8 active XMM registers used in the inner loop, so there will likely be some stack shuffling offsetting the SSE gains.

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.

**Pull request closed without merging**
